src/inspector_agent.cc

```diff
@@ -0,0 +1,515 @@
+  auto env = parent_env_;
+  inspector_ = new V8NodeInspector(this, env, platform);
+
+  int err;
+
+  platform_ = platform;
+
+  err = uv_loop_init(&child_loop_);
+  CHECK_EQ(err, 0);
+  err = uv_async_init(env->event_loop(), &data_written_, nullptr);
+  CHECK_EQ(err, 0);
+  err = uv_mutex_init(&queue_lock_);
+  CHECK_EQ(err, 0);
+  err = uv_mutex_init(&pause_lock_);
+  CHECK_EQ(err, 0);
+  err = uv_cond_init(&pause_cond_);
+  CHECK_EQ(err, 0);
+
+  uv_unref(reinterpret_cast<uv_handle_t*>(&data_written_));
+
+  port_ = port;
+  wait_ = wait;
+
+  err = uv_thread_create(&thread_,
+                         reinterpret_cast<uv_thread_cb>(Agent::ThreadCbIO),
+                         this);
+  CHECK_EQ(err, 0);
+  uv_sem_wait(&start_sem_);
+
+  if (wait) {
+    // Flush messages in case of wait to connect, see OnRemoteDataIO on how it
+    // should be fixed.
+    SetConnected(true);
+    PostMessages();
+  }
+}
+
+void Agent::Stop() {
+  // TODO(repenaxa): hop on the right thread.
+  DisconnectAndDisposeIO(client_socket_);
+  int err = uv_thread_join(&thread_);
+  CHECK_EQ(err, 0);
+
+  uv_run(&child_loop_, UV_RUN_NOWAIT);
+
+  err = uv_loop_close(&child_loop_);
+  CHECK_EQ(err, 0);
+  delete inspector_;
+}
+
+bool Agent::IsStarted() {
+  return !!platform_;
+}
+
+void Agent::WaitForDisconnect() {
+  inspector_->runMessageLoopOnPause(0);
+}
+
+// static
+void Agent::ThreadCbIO(Agent* agent) {
+  agent->WorkerRunIO();
+}
+
+// static
+void Agent::OnSocketConnectionIO(uv_stream_t* server, int status) {
+  if (status == 0) {
+    inspector_socket_t* socket = reinterpret_cast<inspector_socket_t*>(
+        malloc(sizeof(inspector_socket_t)));
+    ASSERT_NE(nullptr, socket);
+    memset(socket, 0, sizeof(inspector_socket_t));
+    socket->data = server->data;
+    if (inspector_accept(server, socket, Agent::OnInspectorHandshakeIO) != 0) {
+      free(socket);
+    }
+  }
+}
+
+// static
+bool Agent::OnInspectorHandshakeIO(inspector_socket_t* socket,
+                                 enum inspector_handshake_event state,
+                                 const char* path) {
+  Agent* agent = reinterpret_cast<Agent*>(socket->data);
+  switch (state) {
+  case kInspectorHandshakeHttpGet:
+    return RespondToGet(socket, path);
+  case kInspectorHandshakeUpgrading:
+    return AcceptsConnection(socket, path);
+  case kInspectorHandshakeUpgraded:
+    agent->OnInspectorConnectionIO(socket);
+    return true;
+  case kInspectorHandshakeFailed:
+    return false;
+  default:
+    ASSERT(false);
+  }
+}
+
+// static
+void Agent::OnRemoteDataIO(uv_stream_t* stream,
+                           ssize_t read,
+                           const uv_buf_t* b) {
+  inspector_socket_t* socket =
+      reinterpret_cast<inspector_socket_t*>(stream->data);
+  Agent* agent = reinterpret_cast<Agent*>(socket->data);
+  if (read > 0) {
+    uv_mutex_lock(&agent->queue_lock_);
+    blink::protocol::String16 str(b->base, read - 1);
```

Comment:
It is an artifact from an older implementation (that had terminating '\0') - I will fix it.